aptos-move/block-executor/src/view.rs

```diff
@@ -545,10 +534,45 @@ impl<'a, T: Transaction, X: Executable> ParallelState<'a, T, X> {
 
         loop {
             match self.versioned_map.data().fetch_data(key, txn_idx) {
-                Ok(Versioned(version, v, layout)) => {
-                    let data_read = DataRead::Versioned(version, v.clone(), layout)
+                Ok(Versioned(version, value)) => {
+                    // If we have a known layout, upgrade RawFromStorage value to Exchanged.
+                    if let UnknownOrLayout::Known(layout) = layout {
+                        if let ValueWithLayout::RawFromStorage(v) = value {
+                            match patch_base_value(v.as_ref(), layout) {
+                                Ok(patched_value) => {
+                                    self.versioned_map.data().set_base_value(
+                                        key.clone(),
+                                        ValueWithLayout::Exchanged(
+                                            Arc::new(patched_value),
+                                            layout.cloned().map(Arc::new),
+                                        ),
+                                    );
+                                    // Refetch in case a concurrent change went through.
+                                    continue;
+                                },
+                                Err(e) => {
+                                    error!("Couldn't patch value from versioned map: {}", e);
+                                    self.captured_reads.borrow_mut().mark_incorrect_use();
+                                    return ReadResult::HaltSpeculativeExecution(
+                                        "Couldn't patch value from versioned map".to_string(),
+                                    );
+                                },
+                            }
+                        }
+                    }
+
```

Comment:
So the value can be RawFromStorage iff layout is Unknown. Can we assert it?

Reply:
that's exactly what above ifs do - but we can assert here as well (and that can be just an assert, not raising Err, as invariant is locally verifiable).